### PR TITLE
[kilppa] support subsetting OS/2 table

### PR DESCRIPTION
**Pull request by qxliu76**

### CODE REVIEW

File: klippa/src/lib.rs [hunk 1]
```diff
@@ -461,6 +464,10 @@ fn subset_table<'a>(
         Maxp::TAG => {
             subset_maxp(font, plan, builder)?;
         }
+
+        Os2::TAG => {
+            subset_os2(font, plan, builder)?;
+        }
```
Comment (rsheeter) [inline klippa/src/lib.rs:470]: Here and above ther eseem to be a lot of extra {}, why not `Os2::TAG => subset_os2(font, plan, builder)?,`

File: klippa/src/os2.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    }
+}
+
+// Returns the bit to be set in os/2 ulUnicodeOS2Range for a given codepoint.
+fn get_unicode_range_bit(cp: u32) -> Option<u8> {
+    match OS2_UNICODE_RANGES.binary_search_by(|&(a, b, _)| OS2Range::new(a, b).cmp(cp)) {
+        Ok(i) => Some(OS2_UNICODE_RANGES[i].2),
+        Err(_) => None,
+    }
```
Comment (cmyr) [inline klippa/src/os2.rs:82]: you can use `Result::ok` to convert a result to an option, like
```suggestion
    OS2_UNICODE_RANGES.binary_search_by(|&(a, b, _)| OS2Range::new(a, b).cmp(cp))
        .ok()
        .map(|i| OS2_UNICODE_RANGES[i].2)
```

**Review by cmyr — APPROVED** [review]
looks good!

File: klippa/src/lib.rs (hunk 1 above)
Comment (rsheeter) [reply klippa/src/lib.rs:470]: Most of your branches have your functions return type. It looks like you could drop the final `Ok(())` and write this function as just the match, e.g.

```rust
fn subset_table<'a>(
    tag: Tag,
    font: &FontRef<'a>,
    plan: &Plan,
    builder: &mut FontBuilder<'a>,
) -> Result<(), SubsetError> {
    match tag {
        Glyf::TAG => subset_glyf_loca(plan, font, builder),
        // head is handled by glyf table if exists
        Head::TAG => font.glyf().map(|_| ()).or_else(|_| subset_head(font, plan, builder)),            
        //Skip, handled by Hmtx
        Hhea::TAG => Ok(()),
        Hmtx::TAG => subset_hmtx_hhea(font, plan, builder),
        //Skip, handled by glyf
        Loca::TAG => Ok(()),
        Maxp::TAG => subset_maxp(font, plan, builder),
        Os2::TAG => subset_os2(font, plan, builder),
        _ => if let Some(data) = font.data_for_tag(tag) {
                builder.add_raw(tag, data);
                Ok(())
            } else {
                Err(SubsetError::SubsetTableError(tag))
            }
    }
}
```

Comment (qxliu76) [reply klippa/src/lib.rs]: Thanks!

File: klippa/src/os2.rs (hunk 2 above)
Comment (qxliu76) [reply klippa/src/os2.rs]: thanks!

File: klippa/src/os2.rs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+use write_fonts::{
+    read::{
+        collections::IntSet,
+        tables::os2::{Os2, OS2_UNICODE_RANGES},
+        FontRef, TableProvider, TopLevelTable,
+    },
+    FontBuilder,
+};
+
+pub fn subset_os2(
```
Comment (rsheeter) [inline klippa/src/os2.rs:14]: Would pub(crate) better reflect the intent?

Comment (rsheeter) [inline klippa/src/os2.rs:14]: Is there equivalent harfbuzz and/or fonttools code we could reference? If we're not doing that for our subset fns perhaps we could start? - doing it for freetype has been quite helpful.

File: klippa/src/os2.rs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        .unwrap()
+        .copy_from_slice(&us_last_char_index.to_be_bytes());
+
+    if plan
+        .subset_flags
+        .contains(SubsetFlags::SUBSET_FLAGS_NO_PRUNE_UNICODE_RANGES)
+    {
+        builder.add_raw(Os2::TAG, out);
+        return Ok(());
+    }
```
Comment (rsheeter) [inline klippa/src/os2.rs:38]: If anyone adds code below this they expect to execute the return may surprise them, plus you duplicate adding to the builder. Could you do it the other way around, as in if plan doesn't contain no prune update ranges, no early return?

**Comment by qxliu76** [comment]
> Question, where did all the test data come from? Are we copying it from harfbuzz?

yes, they're copied from harfbuzz, but I tailored them to drop those unsupported tables

**Review by garretrieger — APPROVED** [review]

**Pull request merged**